File: backend/core/mp_date_validation.py
"""Validación de rangos de fecha para consultas a Mercado Público (escaneo día a día)."""

from datetime import datetime, timedelta

MP_MAX_RANGE_DAYS = 31
MP_MAX_YEARS_BACK = 2


def validate_mp_date_range(fecha_inicio: str, fecha_fin: str) -> dict:
    """
    Returns:
        dict con keys: valid (bool), error (str|None), day_count (int|None)
    """
    if not fecha_inicio or not fecha_fin:
        return {'valid': False, 'error': 'Debe indicar fecha_inicio y fecha_fin.'}

    try:
        start_dt = datetime.strptime(fecha_inicio, '%Y-%m-%d').date()
        end_dt = datetime.strptime(fecha_fin, '%Y-%m-%d').date()
    except ValueError:
        return {'valid': False, 'error': 'Formato de fecha inválido. Use YYYY-MM-DD.'}

    today = datetime.now().date()

    if end_dt < start_dt:
        return {'valid': False, 'error': 'fecha_fin debe ser igual o posterior a fecha_inicio.'}

    if end_dt > today:
        return {'valid': False, 'error': 'fecha_fin no puede ser posterior a hoy.'}

    min_start = today - timedelta(days=MP_MAX_YEARS_BACK * 365)
    if start_dt < min_start:
        return {
            'valid': False,
            'error': f'fecha_inicio no puede ser anterior a {MP_MAX_YEARS_BACK} años.',
        }

    day_count = (end_dt - start_dt).days + 1

    if day_count > MP_MAX_RANGE_DAYS:
        return {
            'valid': False,
            'error': (
                f'El período máximo es {MP_MAX_RANGE_DAYS} días '
                f'({day_count} solicitados). Use un rango menor o búsqueda por código.'
            ),
            'day_count': day_count,
        }

    return {'valid': True, 'day_count': day_count, 'error': None}

File: frontend/src/utils/mpDateValidation.js
export const MP_MAX_RANGE_DAYS = 31;
export const MP_WARN_RANGE_DAYS = 15;
export const MP_MAX_YEARS_BACK = 2;

const parseLocalDate = (iso) => {
    const [y, m, d] = iso.split('-').map(Number);
    return new Date(y, m - 1, d);
};

const todayIso = () => {
    const t = new Date();
    const y = t.getFullYear();
    const m = String(t.getMonth() + 1).padStart(2, '0');
    const d = String(t.getDate()).padStart(2, '0');
    return `${y}-${m}-${d}`;
};

/**
 * Valida rango fecha_inicio / fecha_fin para consultas Mercado Público (día a día).
 * @returns {{ valid: boolean, error?: string, warning?: string, dayCount?: number }}
 */
export const validateMpDateRange = (fechaInicio, fechaFin) => {
    if (!fechaInicio || !fechaFin) {
        return { valid: false, error: 'Debe indicar fecha de inicio y fecha de fin.' };
    }

    const start = parseLocalDate(fechaInicio);
    const end = parseLocalDate(fechaFin);
    const today = parseLocalDate(todayIso());

    if (Number.isNaN(start.getTime()) || Number.isNaN(end.getTime())) {
        return { valid: false, error: 'Las fechas ingresadas no son válidas.' };
    }

    if (end < start) {
        return { valid: false, error: 'La fecha de fin debe ser igual o posterior a la de inicio.' };
    }

    if (end > today) {
        return { valid: false, error: 'La fecha de fin no puede ser posterior a hoy.' };
    }

    const minStart = new Date(today);
    minStart.setFullYear(minStart.getFullYear() - MP_MAX_YEARS_BACK);
    if (start < minStart) {
        return { valid: false, error: `La fecha de inicio no puede ser anterior a ${MP_MAX_YEARS_BACK} años.` };
    }

    const dayCount = Math.floor((end - start) / (24 * 60 * 60 * 1000)) + 1;

    if (dayCount > MP_MAX_RANGE_DAYS) {
        return {
            valid: false,
            error: `El período máximo es ${MP_MAX_RANGE_DAYS} días (${dayCount} seleccionados). Sincronice por semanas o use búsqueda por código.`,
            dayCount,
        };
    }

    let warning;
    if (dayCount >= MP_WARN_RANGE_DAYS) {
        warning = `Consultará ${dayCount} días (Mercado Público responde día a día). Puede tardar varios minutos.`;
    }

    return { valid: true, warning, dayCount };
};

export const validateMpCodeSearch = (code) => {
    const trimmed = (code || '').trim();
    if (!trimmed) {
        return { valid: false, error: 'Ingrese un código de documento válido.' };
    }
    return { valid: true, code: trimmed };
};
